src/main/kotlin/org/jmailen/gradle/kotlinter/tasks/format/FormatWorkerAction.kt

```diff
@@ -26,29 +25,33 @@ abstract class FormatWorkerAction : WorkAction<FormatWorkerParameters> {
     private val output: File? = parameters.output.asFile.orNull
 
     override fun execute() {
-        resetEditorconfigCacheIfNeeded(
+        val ktLintEngine = KtLintRuleEngine(
+            ruleProviders = resolveRuleProviders(defaultRuleSetProviders, ktLintParams.experimentalRules),
+            editorConfigOverride = editorConfigOverride(ktLintParams),
+        )
+
+        ktLintEngine.resetEditorconfigCacheIfNeeded(
             changedEditorconfigFiles = parameters.changedEditorConfigFiles,
             logger = logger,
         )
 
         val fixes = mutableListOf<String>()
         try {
             files.forEach { file ->
-                val ruleSets = resolveRuleProviders(defaultRuleSetProviders, ktLintParams.experimentalRules)
```

Comment:
I wasn't sure why rulesets were resolved for each file, was there a reason for that? Did I break something by moving rules resolution out of the loop?

Reply:
>I wasn't sure why rulesets were resolved for each file, was there a reason for that?

I'm not aware of any 🤷 I agree the change you proposed should work just fine 👀

Reply:
You know I have a vague memory that they were stateful and if you didn't re-resolve them in the action you'd run into problems. Superstitiously I might recommend keeping the behavior. If you want to change it, make sure to run the PR version on a real project with a number of files and source sets.

Reply:
Hah, based on the documentation it looks like you're right in that the rules might be stateful, but at the same time `RuleSetProviderV2` solves statefulness issues (provided it's implemented as intended) and it's explicitly said that the engine won't reuse rules:

```kotlin
/**
 * The [Rule] contains the life cycle hooks which are needed for the KtLint rule engine.
 *
 * Implementation **doesn't** have to be thread-safe or stateless (provided that [RuleSetProviderV2] creates a new
 * instance of the [Rule] when [RuleSetProviderV2.getRuleProviders] has implemented its [RuleProvider] in such a way
 * that each call to [RuleProvider.createNewRuleInstance] indeed creates a new instance. The KtLint engine never
 * re-uses a [Rule] instance once is has been used for traversal of the AST of a file.
 */
public open class Rule(
```

It might be a good case to solve when adding tests involving custom rulesets though. I'll also check the current implementation on my $workProject 👍

Reply:
Yeah, they may well have fixed this issue in later ktlint versions -- so long as it tests out ok, I'm good with a change here 👍

Reply:
Actually, discovered you do in fact need to resolve rule providers for each file. If you don't, you get an error when KtLint tries to parse the .editorconfig on rule like
```
ktlint_standard_no-wildcard-imports = disabled
```

It will complain that ruleSetExecution is null in the VisitorProvider.

I'm putting together a 3.14.0 release and I'll fix there.